mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,621 @@
+        # partition along last axis by default
+        partition_ = np.partition(preds_single, len(preds_single[0]) - 2)
+        # diff between the highest two values
+        confidences = np.abs(partition_[:, -1] - partition_[:, -2])
+        idxs = custom_reordering(
+            confidences, do_rank=do_rank, sampling_size=sampling_size
+        )
+        sampled_indices, unsampled_indices, ranked_indices = (
+            idxs[:sampling_size].tolist(),
+            idxs[sampling_size:].tolist(),
+            idxs.tolist() if do_rank else None,
+        )
+        return sampled_indices, unsampled_indices, ranked_indices
+
+
+class EntropySampling(Heuristic):
+    """ Selection strategy that select queries with the highest entropy."""
+
+    def _extractor(
+        self, sampling_size: int, preds_single: List[List[float]], do_rank: bool = True, **kwargs
+    ):
+        """
+        Args:
+            sampling_size (int): Size of data being sampled in a turn.
+            preds_single: is a list[list[float]] and contains probability scores for each data
+                point for each class.
+            do_rank: if True, returns a ranked list of the indices for confidence-sorting of data
+                samples.
+        """
+        sampling_size = Heuristic.validate_sampling_size(sampling_size)
+        entropies = scipy_entropy(np.array(preds_single), axis=1, base=ENTROPY_LOG_BASE)
+        confidences = -entropies
+        idxs = custom_reordering(
+            confidences, do_rank=do_rank, sampling_size=sampling_size
+        )
+        sampled_indices, unsampled_indices, ranked_indices = (
+            idxs[:sampling_size].tolist(),
+            idxs[sampling_size:].tolist(),
+            idxs.tolist() if do_rank else None,
+        )
+        return sampled_indices, unsampled_indices, ranked_indices
+
+
+class DisagreementSampling(Heuristic):
+    """ Selection strategy that measures 'disagreement' across multiple classifiers."""
+
+    def _extractor(
+        self,
+        sampling_size: int,
+        preds_multi: List[List[List[float]]],
+        do_rank: bool = True,
+        **kwargs,
+    ):
+        """
+        Args:
+            sampling_size (int): Size of data being sampled in a turn.
+            preds_multi: is a list of list[list[float]] and contains probability scores for each
+                data point for each class.
+            do_rank: if True, returns a ranked list of the indices for confidence-sorting of data
+                samples.
+        """
+        sampling_size = Heuristic.validate_sampling_size(sampling_size)
+        choices = []
+        for pred in preds_multi:
+            model_choices = np.argmax(pred, axis=1).tolist()
+            choices.append(model_choices)
+        # Transpose to make rows correspond to datapoints, and columns to models
+        choices = np.array(choices).T
+        confidences = [
+            Counter(row.tolist()).most_common(1)[0][1] / len(row) for row in choices
```

Comment:
Its unclear what `Counter(row.tolist()).most_common(1)[0][1]` represents

Reply:
Refactored, this part of the code was obtaining the frequency of the most frequently selected class by the classifiers. Refactored version:


```
        for row in chosen_classes_per_element:
            class_counts_per_element = Counter(row)
            _, freq = class_counts_per_element.most_common()[0]
            num_models = len(row)
            percent_voted_most_req_class = freq / num_models
            disagreement_score = 1 - percent_voted_most_req_class
            disagreement_scores.append(disagreement_score)
```